dali/operators/generic/one_hot.h

```diff
@@ -31,17 +31,24 @@ namespace detail {
 template<typename Out, typename In>
 void DoOneHot(kernels::OutTensorCPU<Out, DynamicDimensions> output,
               kernels::InTensorCPU<In, DynamicDimensions> input, int num_classes,
-              same_as_t<Out> on_value, same_as_t<Out> off_value, int ndims, int axis = 0) {
+              same_as_t<Out> on_value, same_as_t<Out> off_value, int axis) {
   auto in = input.data;
   auto out = output.data;
-  for (int i = 0; i < volume(output.shape); ++i) out[i] = off_value;
-  int64_t inner = volume(input.shape.last(ndims - axis));
-  for (int64_t i = 0, n = input.num_elements(); i < n; i++) {
-    int cls = in[i];
-    if (cls < 0 || cls >= num_classes)
-      continue;
-    int64_t out_idx = (i % inner) + cls * inner + (i / inner) * (inner * num_classes);
-    out[out_idx] = on_value;
+  axis = axis < 0 ? output.shape.sample_dim() - 1 : axis;
+  auto outer = output.shape.first(axis);
+  auto inner = output.shape.last(output.shape.sample_dim() - axis - 1);
+  auto volume_outer = volume(outer);
+  auto volume_inner = volume(inner);
```

Comment:
We don't need to create the intermediate shapes if their only purpose is to calculate the volume.
```suggestion
  auto volume_outer = volume(outer.begin(), outer.begin() + axis);
  auto volume_inner = volume(outer.begin() + axis + 1, outer.ebd());
```
But won't push it - if you find current solution to be more readable, then keep it.

Reply:
Done, forgot about that overload of volume tbh.